src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+            for(int a : sq)
+                cur_pair[1][a] = -1;
+        }
+
+        return op;
+    }
+
+    bool is_transpose_identity(int_vec perm) const
+    {
+        for(auto i = 0u; i < perm.size(); ++i)
+            if(perm[i] != i)
+                return false;
+
+        return true;
+    }
+
+    instruction_ref apply_transpose_op(const onnx_parser::node_info& info,
+                                       instruction_ref op,
+                                       const int_vec& perm,
+                                       int_vec& row) const
+    {
+        if(is_transpose_identity(perm))
+            return op;
+
+        op = info.add_instruction(make_op("transpose", {{"permutation", perm}}), op);
+        // compute output row
+        auto cpy = row;
+        for(auto i = 0u; i < perm.size(); ++i)
+            row[i] = cpy[perm[i]];
+
+        return op;
+    }
+
+    std::pair<instruction_ref, instruction_ref>
+    apply_broadcast_op(const onnx_parser::node_info& info,
+                       instruction_ref opl,
+                       instruction_ref opr,
+                       const int_vec& common_labels) const
+    {
+        std::pair<instruction_ref, instruction_ref> ret;
+
+        auto llens = opl->get_shape().lens();
+        auto rlens = opr->get_shape().lens();
+
+        bool lbc = false;
+        bool rbc = false;
+        for(auto l : common_labels)
+        {
+            if(llens[l] == 1 and rlens[l] == 1)
+                continue;
+
+            if(llens[l] == 1)
+            {
+                lbc      = true;
+                llens[l] = rlens[l];
+            }
+
+            if(rlens[l] == 1)
+            {
+                rbc      = true;
+                rlens[l] = llens[l];
+            }
+        }
+
+        if(lbc)
+            opl = info.add_instruction(make_op("multibroadcast", {{"out_lens", llens}}), opl);
+        if(rbc)
+            opr = info.add_instruction(make_op("multibroadcast", {{"out_lens", rlens}}), opr);
+
+        ret.first  = opl;
+        ret.second = opr;
+        return ret;
+    }
+
+    instruction_ref apply_reduce_sum_op(const onnx_parser::node_info& info,
+                                        instruction_ref op,
+                                        const int_vec& axes,
+                                        int_vec& row) const
+    {
+        if(axes.empty())
+            return op;
+
+        for(int a : axes)
+            row[a] = -1;
+
+        return info.add_instruction(make_op("reduce_sum", {{"axes", axes}}), op);
+    }
+
+    // Utility
+
+    int_mat make_matrix(int cur_pair, int cols, int fill_value) const
+    {
+        return int_mat(cur_pair, int_vec(cols, fill_value));
+    }
+
+    int_vec extract_column(int_mat map_mat, int col_idx, int row_begin, int row_end) const
+    {
+        int_vec ret;
+        ret.reserve(row_end - row_begin);
+
+        for(int i = row_begin; i < row_end; ++i)
+            ret.push_back(map_mat[i][col_idx]);
```

Comment:
This can be written using `std::transform`:

```cpp
std::transform(map_mat.begin()+row_begin, map_mat.begin()+row_end, std::back_inserter(ret), [&](auto& x) {
    return x[col_idx];
});
```

Reply:
Done.